lang/cs/Org.Apache.REEF.Network/Elastic/Task/Workflow.cs

```diff
@@ -0,0 +1,298 @@
+using Org.Apache.REEF.Tang.Annotations;
+using Org.Apache.REEF.Tang.Exceptions;
+using Org.Apache.REEF.Utilities.Attributes;
+using Org.Apache.REEF.Utilities.Logging;
+using System;
+using System.Collections;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading;
+
+namespace Org.Apache.REEF.Network.Elastic.Task
+{
+    /// <summary>
+    /// Task-side representation of the the sequence of group communication operations to execute.
+    /// Exception rised during execution are managed by the framework and recovered through the user-defined
+    /// policies / mechanisms.
+    /// </summary>
+    [Unstable("0.16", "API may change")]
+    public sealed class Workflow : IEnumerator<IElasticOperator>
+    {
+        private static readonly Logger LOGGER = Logger.GetLogger(typeof(Workflow));
+
+        private int _position = -1;
+        private bool _failed;
+        private bool _disposed;
+        private List<int> _iteratorsPosition;
+
+        private readonly object _lock;
+        private readonly IList<IElasticOperator> _operators;
+        private readonly CancellationSource _cancellationSource;
+        private readonly bool _isRescheduled;
+
+        /// <summary>
+        /// Injectable constructor.
+        /// </summary>
+        /// <param name="cancellationSource"></param>
+        [Inject]
+        private Workflow(
+            [Parameter(typeof(GroupCommunicationConfigurationOptions.IsRescheduled))] bool isRescheduled,
+            CancellationSource cancellationSource)
+        {
+            _operators = new List<IElasticOperator>();
+            _failed = false;
+            _disposed = false;
+            _lock = new object();
+            _iteratorsPosition = new List<int>();
+            _cancellationSource = cancellationSource;
+            _isRescheduled = isRescheduled;
+        }
+
+        /// <summary>
+        /// The current iteration value.
+        /// </summary>
+        public object Iteration
+        {
+            get
+            {
+                if (_iteratorsPosition.Count == 0)
+                {
+                    return 0;
+                }
+                else
+                {
+                    var iterPos = _iteratorsPosition[0];
+                    var iterator = _operators[iterPos] as IElasticIterator;
+                    return iterator.Current;
+                }
+            }
+        }
+
+        /// <summary>
+        /// Try to move to the next operation in the workflow.
+        /// </summary>
+        /// <returns></returns>
+        public bool MoveNext()
+        {
+            _position++;
+
+            if (_failed || _cancellationSource.IsCancelled)
+            {
+                return false;
+            }
+
+            // Check if we need to iterate
+            if (_iteratorsPosition.Count > 0 && _position == _iteratorsPosition[0])
+            {
+                var iteratorOperator = _operators[_position] as IElasticIterator;
```

Comment:
I don't like that cast. Can we enforce that `_operators` is a list of `IElasticIterator` elements instead?

Reply:
What about this:
```cs              
IElasticIterator iteratorOperator;                
switch(_operators[_position])
{
    case IElasticIterator iter:
        iteratorOperator = iter;
        break;
     default:
         throw new IllegalStateException("Operator not Iterator");
         break;
}
```